sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +202,390 @@ protected AttestationToken()
+                }
+                return null;
+            }
+        }
 
         /// <summary>
         /// Time at which this token was issued.
         /// </summary>
-        public DateTimeOffset IssuedAtTime { get => DateTimeOffset.FromUnixTimeSeconds(Payload.IssuedAtTime); }
+        public virtual DateTimeOffset? IssuedAtTime
+        {
+            get
+            {
+                if (Payload.IssuedAtTime.HasValue)
+                {
+                    return DateTimeOffset.FromUnixTimeSeconds((long)Payload.IssuedAtTime.Value);
+                }
+                return null;
+            }
+        }
 
         /// <summary>
         /// Represents the body of the token encoded as a string.
         /// </summary>
-        public string TokenBody  { get => Encoding.UTF8.GetString(TokenBodyBytes.ToArray()); }
+        public virtual string TokenBody  { get => Encoding.UTF8.GetString(TokenBodyBytes.ToArray()); }
 
         /// <summary>
         /// Represents the body of the token encoded as a string.
         /// </summary>
-        public string TokenHeader { get => Encoding.UTF8.GetString(TokenHeaderBytes.ToArray()); }
+        public virtual string TokenHeader { get => Encoding.UTF8.GetString(TokenHeaderBytes.ToArray()); }
 
         /// <summary>
         /// Validate a JSON Web Token returned by the MAA.
+        /// <para/>
+        /// If the caller provides a set of signers, than that set of signers will be used as the complete set of candidates for signing.
+        /// If the caller does not provide a set of signers, then the <see cref="ValidateTokenAsync(TokenValidationOptions, IReadOnlyList{AttestationSigner}, CancellationToken)"/>
+        /// API will a set of callers derived from the contents of the attestation token.
         /// </summary>
+        /// <param name="options">Options used while validating the attestation token.</param>
         /// <param name="attestationSigningCertificates">Signing Certificates used to validate the token.</param>
-        /// <param name="validationCallback">User provided callback which allows the customer to validate the token.</param>
-        /// <returns></returns>
-        public virtual bool ValidateToken(IReadOnlyList<AttestationSigner> attestationSigningCertificates, Func<AttestationToken, AttestationSigner, bool> validationCallback = default)
+        /// <param name="cancellationToken">Token used to cancel this operation if necessary.</param>
+        /// <returns>true if the token was valid, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if the signing certificates provided are invalid.</exception>
+        /// <exception cref="Exception">Thrown if validation fails.</exception>
+        public virtual async Task<bool> ValidateTokenAsync(TokenValidationOptions options, IReadOnlyList<AttestationSigner> attestationSigningCertificates, CancellationToken cancellationToken = default)
```

Comment:
This comment is for both `ValidateTokenAsync` and `ValidateTokenSync`.  Why is the cancellation token only checked one time -- in the implementation of `GetCandidateSigningCertificates`?  Does it truly have a need and value?  If it does, shouldn't it be checked more often?

Reply:
But the pattern is that public APIs that can be asynchronous come in two forms: A synchronous api and an async API, both of which take cancellation tokens. The only reason the token is checked here is to avoid a compiler diagnostic.

At some point, this code will likely have logic to retrieve values off-node (JKU/XCU support). At that point, the async functionality will become critical.